elasticapm/instrumentation/packages/httplib2.py

```diff
@@ -0,0 +1,100 @@
+#  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+#  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from elasticapm.conf import constants
+from elasticapm.instrumentation.packages.base import AbstractInstrumentedModule
+from elasticapm.traces import DroppedSpan, capture_span, execution_context
+from elasticapm.utils import get_host_from_url, sanitize_url, url_to_destination
+from elasticapm.utils.disttracing import TracingOptions
+
+
+class Httplib2Instrumentation(AbstractInstrumentedModule):
+    name = "httplib2"
+
+    instrument_list = [("httplib2", "Http.request")]
+
+    # https://httplib2.readthedocs.io/en/latest/libhttplib2.html#httplib2.Http.request
+    args_list = ("url", "method", "body", "headers")
+
+    def call(self, module, method, wrapped, instance, args, kwargs):
+        # grab agrs passed in as positional
+        params = dict(zip(self.args_list, args))
+
+        # fall back to kwargs for args not supplied as positional
+        for arg in self.args_list:
+            if arg in params:
+                continue
+            if arg in kwargs:
+                params[arg] = kwargs[arg]
+
+        if "headers" in params:
+            headers = params["headers"]
+        else:
+            headers = {}
+            kwargs["headers"] = headers
+
+        signature = params.get("method", "GET").upper()
+        signature += " " + get_host_from_url(params["url"])
+        url = sanitize_url(params["url"])
+        destination = url_to_destination(url)
+        transaction = execution_context.get_transaction()
+        with capture_span(
+            signature,
+            span_type="external",
+            span_subtype="http",
+            extra={"http": {"url": url}, "destination": destination},
+            leaf=True,
+        ) as span:
+            # if httplib2 has been called in a leaf span, this span might be a DroppedSpan.
+            leaf_span = span
+            while isinstance(leaf_span, DroppedSpan):
+                leaf_span = leaf_span.parent
+
+            # It's possible that there are only dropped spans, e.g. if we started dropping spans.
+            # In this case, the transaction.id is used
+            parent_id = leaf_span.id if leaf_span else transaction.id
+            trace_parent = transaction.trace_parent.copy_from(
+                span_id=parent_id, trace_options=TracingOptions(recorded=True)
+            )
+            self._set_disttracing_headers(headers, trace_parent, transaction)
+
+            response, content = wrapped(*args, **kwargs)
+            if span.context:
+                span.context["http"]["status_code"] = response.status
+            span.set_success() if response.status < 400 else span.set_failure()
+            return response, content
+
```

Comment:
We also need to re-implement `mutate_unsampled_call_args`, something like https://github.com/elastic/apm-agent-python/blob/15278433c97d9c81c5ad1dcdf57d27da847be912/elasticapm/instrumentation/packages/urllib3.py#L105-L116

This sets the distributed tracing headers even if this transaction isn't sampled. That way, we can avoid breaking distributed tracing chains if we, for some reason, don't sample the transaction.

Reply:
I saw this code and I was wondering - wouldn't this only set the header if a user is passing in the `headers` kwarg (line 110)? This might work for some libraries where we instrument an "internal" method, but for `httplib2` the request is the public API. Is that correct thought process or am I missing something?

Reply:
@HonzaKral yes, looks like you're right, if the user creates a `HTTP(S)ConnectionPool` directly, they could call `urlopen` with `headers` as a positional argument.

I think we're saved here by the fact that most people probably use a `PoolManager` as shown in the [user guide](https://urllib3.readthedocs.io/en/latest/user-guide.html).

I'll open a PR with a fix for the urllib3 instrumentation, thanks for the heads up!

Reply:
>  they could call urlopen with headers as a positional argument

Yes, they could also not specify headers at all actually, or specify `None` as a positional argument for headers which will be a bit trickier to replace. This is a corner case even in the `httplib2` instrumentation I submitted here...

Reply:
@HonzaKral I tried to capture all cases in #1033 (also noticed that headers can also be set on the `HTTPConnectionPool` object, which will be used as a fallback if no headers are set in the `urlopen` call)

Reply:
added in 62188486